### PR TITLE
submission of executable tutorial (after feedback): run TeamCity on Docker

### PR DESCRIPTION
**Pull request by Chen-Zidi**
# Executable-tutorial: Run TeamCity on Docker

## Members

Chen, Zidi: zidi@kth.se, https://github.com/Chen-Zidi

Chen, Sihan: sihanc@kth.se, https://github.com/Spycsh

## Content

TeamCity is a great CI/CD tool and Docker is a open platform for containerizing, which strongly related 
to DevOps.

We have wrote a blog on Dev.to [here](https://dev.to/chenzidi/run-teamcity-on-docker-5f8b). As explained in the
blog, our tutorial can be executed on the web browser by [Play-With-Docker](https://labs.play-with-docker.com/#).
This is a free online tool. Readers need to register to use this tool.
 
## Our target:

|  criteria   | what we do  |
|  ----  | ----  |
| successfully execute all the commands | yes, also in browser |
| run on Linux  | easy to setup and run |
| gives enough background  | comprehensive background  |
| easy to follow  | well documented  |
| contains easter eggs | yes, we have three tips to show interesting things |
| language is correct  | yes |
| The tutorial is successful (attracts comments and success)  | yes* |

* We tried to make our tutorial published in [play-with-docker classroom](https://training.play-with-docker.com/alacart/).
In order to do that, we created a pull request in the [official repo](https://github.com/play-with-docker/play-with-docker.github.io),  so that the responsible workers could review our tutorial. And they thought our tutorial is useful and they would included in the PWD classroom. You can see the comment from responsible workers from [here](https://github.com/play-with-docker/play-with-docker.github.io/pull/219).
And we get 2 hearts and 2 unicorn in dev.to.

## What we have changed according to feedback

- add more background info in introduction
- move the instruction of PWD and folder creation to step 1
- delete a few explanation in docker run command in step 1
- add wireframes on the figures
- add links to TeamCity document in step 1 tip
- add description that the agent need time to show up in step 4 
- add description that the program need time to complete in step 5
- add link to TeamCity document 
- rewrite conclusion
- add interesting links in conclusion

### CODE REVIEW

**Comment by Chen-Zidi** [comment]
need feedback from @m4reko

**Comment by Chen-Zidi** [comment]
@khaes-kth

**Comment by m4reko** [comment]
Aah, I thought that I was done with school for this week haha 😉

Jokes aside, I will hopefully get back to you with feedback in less than 24 hours. Thanks for the mention.

**Comment by m4reko** [comment]
I did sadly not have time to complete the feedback this weekend but here you go @Chen-Zidi @Spycsh 

# Feedback
Author: Markus Wesslén (mwesslen@kth.se)

## Overall
- This tutorial is nicely put together and the chosen platforms and tools work well together.
- The time estimate is good. I managed to do everything in just under 20 min even though I had to restart one time early on due to some problems with Play-with-Docker.
- The instructions are divided into logical steps and most of the instructions are easy to follow.
- Its good that you have an introduction and some sort of concluding text but those parts can be improved. More on that further down.
- Sometimes it's distracting to manage the three tabs that are needed for this tutorial but I guess that's hard to do anything about with the tools you chose. In [Katacoda](https://www.katacoda.com/), that I used for my tutorial, you can have everything in the same browser window and even have the TeamCity web interface in the same window.

## Sections
### Introduction
- It's nice that you include an introduction but I feel that it falls short in the content. In an introduction I want to be intrigued and get some background on whats included but your intro is purely technical. I, that have never encountered TeamCity before, have no clue what I'm going to learn after your introduction which also makes it harder to grasp the steps the first tie you work through them. Give some background on TeamCity, tell me what its used for and make it interesting so that i want to fulfil the tutorial.
- It good that you describe your tools (like Play-with-Docker) in the intro but I feel like the creation of a new instance belongs in the first step.
- Your instructions on how to copy and paste does not work on my mac but good old `cmd-c` `cmd-v` works fine. Maybe you don't have to include this since it's pretty basic? If it's more complicated on Windows the specify the operating system at least.
- The folder creation also belongs in it's own step according to me.

### Step 1
- This part is very clear and straight to the point which I like.
- The explanation of all the command flags for `docker run` is great to have but maybe a bit verbose. You can always reference the docker manual here if the reader want to know more. It feels like some of the parts are outside the scope of this tutorial.
- It's awesome that you include the commands `docker images` and `docker ps` which makes it easy to check one's work and also gives some more insight on how docker is working.

### Step 2
- This step is also clear and includes a lot of pictures of the GUI which makes it easy to follow. You could maybe have used some sort of graphics software (like Paint or [GIMP](https://www.gimp.org/)) to mark the relevant buttons or fields in the pictures.
- The tips are a nice addition but it would be awesome if you added some resources/links here pointing you to blogs or manuals for further reading. Some quick Google searches led me to [this link](https://www.jetbrains.com/help/teamcity/integrating-teamcity-with-docker.html#Docker+Disk+Space+Cleaner) which could maybe be added in your first tip.

### Step 3
- Here we suddenly introduce an agent and I would have loved some background on why we have Agents connecting to our server and what there respective roles are. After some quick reading I understand that it is good if we want to parallelise our builds. A very short description on this either here or in the introduction would be very nice.
- It's good that you only describe the new command flags here. I think all of the are relevant.

### Step 4
- For me it took some time before the agent showed up in this step after starting the agent in the terminal. This made me a bit confused and it would be nice if you were given a heads up on this in the initial part of this step.
- I also missed pictures of some of the popup dialogs from the GUI in this step which also halted my progress a bit. Pictures, preferably with markings for the important buttons, would be awesome.

### Step 5
- You could specify here also that it takes some time before this part is done even if it's much clearer here that something is happening and that you should probably wait.
- Here you also explains that TeamCity can do more than this which is intriguing but a follow up with a link to some resource would be awesome.

### Conclusion
- It's very nice that you include a concluding part but I feel like the content are lacking here in the same way as in the introduction. You provide a short summary of what we learned which is nice but it feels like repetition from the intro since it's presented in a very straight forward way. Wrapping it in some more descriptive and concluding paragraphs would improve the feeling.
- I would also say that this is a perfect place to put even more interesting resources (which I am sure there are lots of) and a call to action to dive deeper into the subject or something like that.

## Summary
In summary this is a very nice tutorial that is easy to follow and have a very clear goal. It makes use of very nice tools and platforms to present and execute the material which is always a plus. The average developer would have no problems executing all the steps and getting TeamCity to run with Docker.

However, it lacks a bit in the introduction and conclusion which both would need some more background, resources and less technical descriptions. This would also help the user to grasp the rest of the content quicker since the tutorial right now contain very little description on what TeamCity is and how it is used.

Overall, you have done a very nice job on this tutorial and I look forward to see your improvements.

**Comment by Chen-Zidi** [comment]
Thanks to the feedback from @m4reko .  We are now ready to be assessed @khaes-kth .

**Comment by khaes-kth** [comment]
Good. I'll merge your PR soon.

**Comment by Chen-Zidi** [comment]
Hi! @khaes-kth I just noticed that this PR is still not merged yet. If there is any problem, please contact us : )

**Comment by khaes-kth** [comment]
Hi @Chen-Zidi 

No there is no problem. I just left it open to make sure that I will consider your case later while grading it.

**Comment by Chen-Zidi** [comment]
I see. Thanks again for your consideration ; )
